Expose os, compiler, arch in `php -v`

arnaud-lb opened the pull request:
See https://github.com/php/php-src/issues/19256

The bug system requires the output of `php -v` when reporting bugs. Exposing more information in `php -v` can be useful.

`PHP_BUILD_COMPILER` and `PHP_BUILD_ARCH` are always defined in Windows build, or when defined in the environment during the build (https://github.com/php/php-src/pull/5834). In other cases I determine these from compiler macros.

Output examples:
```
PHP 8.5.0-dev (cli) (built: Jul 31 2025 11:18:32) (NTS DEBUG x86_64) (Linux 6.15.8-200.fc42.x86_64 Clang 20.1.8 (Fedora 20.1.8-1.fc42))
Copyright (c) The PHP Group
Zend Engine v4.5.0-dev, Copyright (c) Zend Technologies
    with Zend OPcache v8.5.0-dev, Copyright (c), by Zend Technologies
```

```
PHP 8.5.0-dev (cli) (built: Jul 31 2025 11:29:06) (ZTS DEBUG x64) (Windows NT 10.0 Visual C++ 2022)
Copyright (c) The PHP Group
Zend Engine v4.5.0-dev, Copyright (c) Zend Technologies
    with Zend OPcache v8.5.0-dev, Copyright (c), by Zend Technologies
```

Girgias commented on main/main.c:
```diff
@@ -121,26 +121,36 @@ ZEND_ATTRIBUTE_CONST PHPAPI const char *php_build_provider(void)
 
 PHPAPI char *php_get_version(sapi_module_struct *sapi_module)
 {
+	zend_string *os = php_get_uname('s');
```
Can you not use the `PHP_OS_FAMILY` C macro constant here?

mvorisek commented on main/php_main.h:
```diff
@@ -89,6 +89,26 @@ PHPAPI bool php_tsrm_startup(void);
+
+#ifndef PHP_BUILD_ARCH
+# if defined(__x86_64__)
+#  define PHP_BUILD_ARCH "x86_64"
```
```suggestion
#  define PHP_BUILD_ARCH "x64"
```

remicollet commented:
Can be confusing to report both 
- build-time information (GCC, ...)
- run-time information (name)

This information already in `php -i`

```
System => Linux builder.remirepo.net 6.15.8-100.fc41.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Jul 24 12:25:37 UTC 2025 x86_64
Build Date => Jul 29 2025 15:48:33
Build System => Fedora release 41 (Forty One)
Build Provider => Remi's RPM repository <https://rpms.remirepo.net/> #StandWithUkraine
Compiler => gcc (GCC) 14.3.1 20250523 (Red Hat 14.3.1-1)
```

Notice `PHP_BUILD_COMPILER` can be very long, which is OK in phpinfo, but not in `php -v`  (I use a local patch to only display "GCC" there)

IMHO, `PHP_BUILD_PROVIDER` (Fedora ...) when set is more useful that `uname` output (Linux... ), and both can be long

arnaud-lb commented:
> Can be confusing to report both
> 
>     * build-time information (GCC, ...)
>     * run-time information (from uname)

Agreed. I've updated the branch to use only build-time information.

> This information already in `php -i`

I feel this is less convenient than `php -v` for the purpose of bug reporting. The full output of `php -i` is very large, and may expose too much, so it must be filtered / edited.

> Notice `PHP_BUILD_COMPILER` can be very long, which is OK in phpinfo, but not in `php -v` (I use a local patch to only display "GCC" there)

The compiler version can be useful to analyze bugs, as some features may only be enabled on some versions. This is the primary reason I'm interested in making this change. Just the major version would be fine, however.

> IMHO, `PHP_BUILD_PROVIDER` (Fedora ...) when set is more useful that `uname` output (Linux... ), and both can be long

I've updated the output of `php -v` as follows:

Without PHP_BUILD_PROVIDER:

```
PHP 8.5.0-dev (cli) (built: Aug  4 2025 11:29:59) (NTS DEBUG x64)
Copyright (c) The PHP Group
Built on Linux 6.15.9-201.fc42.x86_64 (GCC 15)
Zend Engine v4.5.0-dev, Copyright (c) Zend Technologies
    with Zend OPcache v8.5.0-dev, Copyright (c), by Zend Technologies
```

With PHP_BUILD_PROVIDER:

```
PHP 8.5.0-dev (cli) (built: Aug  4 2025 11:31:50) (NTS DEBUG x64)
Copyright (c) The PHP Group
Built by Fedora Project (Clang 20)
Zend Engine v4.5.0-dev, Copyright (c) Zend Technologies
    with Zend OPcache v8.5.0-dev, Copyright (c), by Zend Technologies
```

remicollet commented:
> With PHP_BUILD_PROVIDER:

Sorry, I was wrong in my previous comment, I think we need to use `PHP_BUILD_SYSTEM` (ie "Fedora release 42") not `PHP_BUILD_PROVIDER` (Fedora project)

Provider is already part on php -v
So a new line for build system, ex:

```
PHP 8.5.0-dev (cli) (built: Aug  4 2025 11:31:50) (NTS DEBUG x64)
Copyright (c) The PHP Group
Built on Fedora release 41 (Forty One) using gcc (GCC) 14.3.1 20250523 (Red Hat 14.3.1-1)      <= this is System and Compiler
Built by Fedora Project         <= this is the provider
Zend Engine v4.5.0-dev, Copyright (c) Zend Technologies
    with Zend OPcache v8.5.0-dev, Copyright (c), by Zend Technologies
```

NattyNarwhal commented:
FWIW, at least with users I support, they seem to use `php -v` more than phpinfo/-i for reporting things. For example, I added the build provider bit, because they saw Zend Engine in the output and thought they were using builds provided by Zend. For the sake of our bug tracker, `-v` is a lot more frequently used.
